### PR TITLE
Feature/code refactoring

### PR DESCRIPTION
**Pull request by mershark**
In this task we accomplished the following:
1. Separated main menu logic into a Menu class.
2. Removed the switch method for menu options.
3. Enhanced the main function by using the Menu class.
4. Ensured adherence to SOLID principles, especially the Single Responsibility Principle.

### CODE REVIEW

**Review by topeogunleye — CHANGES_REQUESTED** [review]
Hi @mershark, @RicardoEgas,

Great Job on the progress made on your project so far. There are some issues that you still need to work on to go to the next project but you are almost there!

Try-Again

> ## (Highlights) Good Points: 💪

1. **Clear Role of `main.rb`**:
   - Your `main.rb` perfectly adheres to the requirement that it should only start your code and ask the user for the initial choice of options (or to quit). It doesn't contain any business logic related to the application's inner workings. Well done!

2. **Code Organization**:
   - Your `main` function is clearly defined and does an excellent job of organizing the user flow. From presenting options to the user, capturing their input, and delegating actions to the `App` class, everything is systematically arranged.

3. **User Interaction**:
   - Your program provides a clear and intuitive user interface by listing out all the available options and asking the user to make a choice. This makes for a seamless user experience.


Overall, your `main.rb` file exhibits a great understanding of clean code principles, separation of concerns, and user experience considerations. Keep up the excellent work! 👏

## Required Changes ♻️
### I couldn't use inline comments because no changes were made to the app.rb file:

---

### Code Review for `create_person` Method

Hello 👋,

Great effort on your project! I've reviewed the `create_person` method you've shared and I'd like to provide some feedback in line with the Single Responsibility Principle (SRP) and the provided project requirements.

1. **Single Responsibility Principle**:
   - Your `create_person` method is currently responsible for both getting input from the user and creating a person (either student or teacher). According to the SRP, a method should have only one responsibility.
   - The requirements emphasize that all requests for input and input validation should be in the main file or other specialized classes/methods. This means that the user input collection (`gets.chomp`) should ideally be separated from the logic that processes this input.

2. **Input Validation**:
   - It's commendable that you've handled the scenario where a user might enter an invalid option for creating either a student or a teacher.
   - However, to fully align with the project requirements, it's recommended to have the input validation logic extracted into a separate method or even a separate class.

3. **Suggestions**:
   - Consider extracting the logic for getting user input into separate methods or a separate class entirely. This can help in making your codebase cleaner and adhering to the SRP.
   - Instead of directly creating a student or teacher inside the `create_person` method, you can simply collect the necessary attributes and then pass them to the appropriate constructor, further simplifying the method.

Here's a simplified version:

```ruby
def create_person
  person_type = get_person_type
  age = get_age
  name = get_name

  case person_type
  when '1'
    create_student(age, name)
  when '2'
    create_teacher(age, name)
  else
    puts 'Invalid option. Please select 1 for Student or 2 for Teacher.'
  end
end

def get_person_type
  print 'Do you want to create a Student (1) or a Teacher (2)? [Input the number]: '
  gets.chomp
end

def get_age
  print 'Age: '
  gets.chomp
end

def get_name
  print 'Name: '
  gets.chomp
end
```

In conclusion, remember to always keep methods as concise as possible, ensuring they adhere to the Single Responsibility Principle. This not only makes your code more maintainable but also more scalable in the long run.

Keep up the good work! 🌟


---

### Code Review for `create_student` Method

Hello again 👋,

I've taken a look at the `create_student` method you've shared. Here's my feedback in alignment with the Single Responsibility Principle (SRP) and the provided project requirements:

1. **Single Responsibility Principle (SRP)**:
   - The `create_student` method currently handles both user input collection and student object creation. As per the SRP, it's ideal for methods to handle one responsibility. The logic for collecting user inputs should be separated from the logic to create a student.
   - The project requirements specifically mention that all user input and validation logic should be extracted to the main file or other specialized methods/classes. This means the `gets.chomp` portions inside the `create_student` method should be moved elsewhere.

2. **Suggestions**:
   - Consider moving the user input collection for "Has parent permission?" and "Classroom" into specialized methods for user input collection.
   - Once the user input is collected, you can then pass these values as parameters to the `create_student` method. This ensures that the `create_student` method is solely responsible for creating a student and not mixed with user input responsibilities.

Here's a suggestion:

```ruby
def create_person
  # ... [rest of the code]
  
  if person_type == "1"
    permission = get_parent_permission
    classroom = get_classroom
    create_student(age, name, permission, classroom)
  # ... [rest of the code]
end

def get_parent_permission
  print 'Has parent permission? [Y/N]: '
  gets.chomp.downcase
end

def get_classroom
  print 'Classroom: '
  gets.chomp
end

def create_student(age, name, permission, classroom)
  parent_permission = permission == 'y'
  @people << Student.new(age, classroom, name, parent_permission: parent_permission)
  puts 'Student Created Successfully'
end
```

By following the above approach, you'll have cleaner, more maintainable methods that adhere to the Single Responsibility Principle.

Keep up with your coding journey, you're doing well! ✨

---

### Code Review for `create_rental` Method

Hello again 👋,

I've reviewed the `create_rental` method within the context of the project requirements and the Single Responsibility Principle (SRP). Here's my feedback:

1. **User Input and Validation**:
   - The project specifically mentions that all requests for user input and input validation should be placed either in the main file or in specialized classes/methods. This means the `gets.chomp` portions and their associated validation checks inside the `create_rental` method should ideally be moved elsewhere.
   - Consider extracting the input collection and validation logic for book and person selection into separate methods.

2. **Suggestions**:
   - Split the method into smaller methods that handle distinct functionalities. For instance:
     - `select_book`
     - `select_person`
     - `input_date`
   - Use these helper methods within `create_rental` to collect and validate input, then use the results to proceed with rental creation.

Here's a suggested refactoring:

```ruby
def create_rental
  book_index = select_book
  return unless book_index

  person_index = select_person
  return unless person_index

  date = input_date

  create_rental_with_indexes(date, book_index, person_index)
end

def select_book
  puts 'Select a book from the following list by number'
  display_books
  book_index = gets.chomp.to_i
  return nil if book_index.negative? || book_index >= @books.length
  
  book_index
end

def select_person
  puts 'Select a person from the following list by number (not ID)'
  display_people
  person_index = gets.chomp.to_i
  return nil if person_index.negative? || person_index >= @people.length
  
  person_index
end

def input_date
  print 'Date (YYYY/MM/DD): '
  gets.chomp
end
```

By following the above suggestions, your code will better adhere to the Single Responsibility Principle, making it more maintainable and easier to understand.

You're doing great and are on the right track! Keep iterating and refining your code based on best practices! ✨

---

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- x
- y
- z

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @topeogunleye in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by 1GORDON — APPROVED** [review]
# Status: APPROVED 🟢 

Hi @mershark, and @RicardoEgas ✋

Your project is complete!  There is nothing else to say other than... it's time to merge it: ship it:
Congratulations! :tada: :clap: :clap: :100:

## Highlights!
- Implemented requested changes ✔️
- App runs as expected ✔️

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend considering them as they can improve your code._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread and tag me `@1GORDON` if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
